lib/server.js

```diff
@@ -305,23 +318,54 @@ CoAPServer.prototype._handle = function(packet, rsinfo) {
   request.rsinfo = rsinfo
   response.statusCode = '2.05'
   response._request = request._packet
+  response._cachekey = toCacheKey(rsinfo.address, rsinfo.port, packet)
 
   var block2cache = this._block2Cache;
+  var block2cache_expiry = this._options.block2CacheExpiryMs;
   //inject this function so the response can add an entry to the cache
   response._addCacheEntry = function(key, payload) {
-    debug('add payload to cache, key:', key);
-    block2cache.key = key;
-    block2cache.payload = payload;
+    if (block2cache.hasOwnProperty(key)) {
+      debug('delete old cache entry, key:', key)
+      clearTimeout(block2cache[key].timeoutId)
+      delete block2cache[key]
+    }
+
+    debug('add payload to cache, key:', key)
+    block2cache[key] = {'payload': payload}
+
+    // setup clean-up timer
+    block2cache[key].timeoutId = setTimeout(function(k) {
+      debug('delete expired cache entry, key:', k)
+      delete block2cache[k]
+    }, block2cache_expiry, key)
```

Comment:
Here's debug output for the new unit-test with parallel requests, key-payload pair is readded to cache after each block2 message:
```
CoAP_Server initialized +1ms
  CoAP_Server no cached entry found, emit request to upper layer +4ms
  CoAP_Server add payload to cache, key: +0ms 62636465/127.0.0.1:9071
  CoAP_Server check if packet token is in cache, key: +1ms ad64864f/127.0.0.1:49238
  CoAP_Server no cached entry found, emit request to upper layer +0ms
  CoAP_Server add payload to cache, key: +0ms ad64864f/127.0.0.1:49238
  CoAP_Server check if packet token is in cache, key: +1ms ad64864f/127.0.0.1:49238
  CoAP_Server found cached payload, key: +0ms ad64864f/127.0.0.1:49238
  CoAP_Server reuse old cache entry, key: +0ms ad64864f/127.0.0.1:49238
  CoAP_Server check if packet token is in cache, key: +1ms ad64864f/127.0.0.1:49238
  CoAP_Server found cached payload, key: +0ms ad64864f/127.0.0.1:49238
  CoAP_Server reuse old cache entry, key: +0ms ad64864f/127.0.0.1:49238
  CoAP_Server check if packet token is in cache, key: +50ms 62636465/127.0.0.1:9071
  CoAP_Server found cached payload, key: +0ms 62636465/127.0.0.1:9071
  CoAP_Server reuse old cache entry, key: +0ms 62636465/127.0.0.1:9071
  CoAP_Server check if packet token is in cache, key: +52ms 62636465/127.0.0.1:9071
  CoAP_Server found cached payload, key: +0ms 62636465/127.0.0.1:9071
  CoAP_Server reuse old cache entry, key: +0ms 62636465/127.0.0.1:9071
  CoAP_Server check if packet token is in cache, key: +52ms 62636465/127.0.0.1:9071
  CoAP_Server found cached payload, key: +0ms 62636465/127.0.0.1:9071
  CoAP_Server reuse old cache entry, key: +0ms 62636465/127.0.0.1:9071
    ✓ parallel requests should not clear block2 cache for each other (162ms)
  CoAP_Server clean-up cache expiry timer, key: +2ms 62636465/127.0.0.1:9071
  CoAP_Server clean-up cache expiry timer, key: +0ms ad64864f/127.0.0.1:49238
```